### PR TITLE
internal/civisibility: api refactor and support for telemetry metrics

### PR DESCRIPTION
**Pull request by tonyredondo**
### What does this PR do?

This PR contains a complete refactor of the test api to follow and adopt Go conventions, and a complete support of the CI Visibility Telemetry Metrics.

https://app.datadoghq.com/dashboard/d4a-6h6-i8q/test-visibility-libraries-telemetry?fromUser=true&refresh_mode=paused&tpl_var_language_name%5B0%5D=go&from_ts=1730971500000&to_ts=1730982300000&live=false

### Motivation

We already have telemetry metrics for all other languages, Go was missing this feature.

### Reviewer's Checklist

- [ ] Changed code has unit tests for its functionality at or near 100% coverage.
- [ ] [System-Tests](https://github.com/DataDog/system-tests/) covering this feature have been added and enabled with the va.b.c-dev version tag.
- [ ] There is a benchmark for any new code, or changes to existing code.
- [ ] If this interacts with the agent in a new way, a system test has been added.
- [ ] Add an appropriate team label so this PR gets put in the right place for the release notes.
- [ ] Non-trivial go.mod changes, e.g. adding new modules, are reviewed by @DataDog/dd-trace-go-guild.

Unsure? Have a question? Request a review!

### CODE REVIEW

File: internal/telemetry/telemetry.go [hunk 1]
```diff
@@ -84,6 +84,8 @@ func (c *client) productChange(namespace Namespace, enabled bool) {
 		products.Products.Profiler = ProductDetails{Enabled: enabled}
 	case NamespaceTracers:
 		// Nothing to do
+	case NamespaceCiVisibility:
+		// Nothing to do
```
Comment (eliottness) [inline internal/telemetry/telemetry.go:88]:
```suggestion
	case NamespaceTracers, NamespaceCiVisibility:
		// Nothing to do
```

Other than that: is there no CiVis product in telemetry ? 🤔 That's sad

File: internal/telemetry/client.go [hunk 2]
```diff
@@ -280,7 +280,16 @@ func (c *client) Stop() {
...
+	if internal.BoolEnv("DD_CIVISIBILITY_ENABLED", false) {
+		cnn := make(chan struct{})
+		c.flushWithCallback(func() {
+			cnn <- struct{}{}
+			close(cnn)
+		})
+		<-cnn
+	} else {
+		c.flush()
+	}
```
Comment (eliottness) [inline internal/telemetry/client.go:292]: How about adding a telemetry client option usable in `ApplyOps` like `SyncFlushOnStop()` or equivalent?

File: internal/telemetry/client.go [hunk 3]
```diff
@@ -443,6 +459,10 @@ func (c *client) flush() {
 				log("submission error: %s", err.Error())
 			}
 		}
+
+		if callback != nil {
+			callback()
+		}
 	}()
```
Comment (eliottness) [inline internal/telemetry/client.go:466]: The whole code that is ran async is this one:
```go
        go func() {
		for _, r := range submissions {
			err := r.submit()
			if err != nil {
				log("submission error: %s", err.Error())
			}
		}

		if callback != nil {
			callback()
		}
        }()
```

How about passing a simple boolean to do something like this?
```go
        submit := func() {
		for _, r := range submissions {
			err := r.submit()
			if err != nil {
				log("submission error: %s", err.Error())
			}
		}

		if callback != nil {
			callback()
		}
        }
        if stopSync {
		submit()
        } else {
		go submit()
        }
```

**Review by eliottness — CHANGES_REQUESTED** [review]
I reviewed ONLY the code in `internal/telemetry`. Good overall

Other than that I would put a BIG warning on the performance hit your product could take considering the ongoing rework of the telemetry lib is partly done because of that: you can have a look a the usage of `telemetry.GlobalClient.Count()` and `telemetry.GlobalClient.Record()` all over dd-trace-go to understand that we never really added telemetry metrics anywhere because of the performance impact these functions have.

File: internal/telemetry/telemetry.go (hunk 1 above)
Comment (tonyredondo) [reply internal/telemetry/telemetry.go]: yeah I know, we still need to work on that.

Comment (tonyredondo) [reply internal/telemetry/telemetry.go]: suggestion applied on [9c8f9e7](https://github.com/DataDog/dd-trace-go/pull/2963/commits/9c8f9e7ce3754968b03f8fa750de3b137371a585)

File: internal/telemetry/client.go (hunk 2 above)
Comment (tonyredondo) [reply internal/telemetry/client.go]: makes sense, I did the change in [9c8f9e7](https://github.com/DataDog/dd-trace-go/pull/2963/commits/9c8f9e7ce3754968b03f8fa750de3b137371a585)

File: internal/telemetry/client.go (hunk 3 above)
Comment (tonyredondo) [reply internal/telemetry/client.go]: yes, make sense, I did the change in [9c8f9e7](https://github.com/DataDog/dd-trace-go/pull/2963/commits/9c8f9e7ce3754968b03f8fa750de3b137371a585)

**Comment by tonyredondo** [comment]
> I reviewed ONLY the code in `internal/telemetry`. Good overall
> 
> Other than that I would put a BIG warning on the performance hit your product could take considering the ongoing rework of the telemetry lib is partly done because of that: you can have a look a the usage of `telemetry.GlobalClient.Count()` and `telemetry.GlobalClient.Record()` all over dd-trace-go to understand that we never really added telemetry metrics anywhere because of the performance impact these functions have.

I understand, I think at test level visibility where we don't have a massive concurrency and very hot paths is acceptable at least for the preview and dogfooding of the library. But, for sure we are looking forward on that rework you mention and to migrate this implementation in the future.

**Review by eliottness — APPROVED** [review]

**Review by darccio — APPROVED** [review]

**Pull request merged**